llvm/lib/Target/AIE/AIE2AsmPrinter.cpp

```diff
@@ -70,6 +70,34 @@ bool AIE2AsmPrinter::lowerOperand(const MachineOperand &MO,
   return LowerAIEMachineOperandToMCOperand(MO, MCOp, *this);
 }
 
+void AIE2AsmPrinter::emitInstruction(const MachineInstr *MI) {
+  // PseudoLoopEnd is just here to carry some extra information. No need to
+  // emit anything
+  if (MI->getOpcode() == AIE2::PseudoLoopEnd)
+    return;
+
+  // Check whether there's a PseudoLoopEnd following. If so, emit the
+  // last-bundle-label that it carries to designate this MI as the last bundle.
```

Comment:
Note: The AsmPrinter has a notion of "post-instruction symbol", and it can trigger the emission of a symbol after an instruction is emitted. See:
```
      // If there is a post-instruction symbol, emit a label for it here.
      if (MCSymbol *S = MI.getPostInstrSymbol())
        OutStreamer->emitLabel(S);
```

Maybe we could hook into this mechanism?

Reply:
And even:
```
      // If there is a pre-instruction symbol, emit a label for it here.
      if (MCSymbol *S = MI.getPreInstrSymbol())
        OutStreamer->emitLabel(S);
```
before emitting the instruction

Reply:
Sounds attractive.